Add CompactOnLaunch.

dalinaum opened the pull request:
Addresses #3739.

dalinaum commented:
Please review this @realm/java

Zhuinden commented on CHANGELOG.md:
```diff
@@ -9,6 +9,8 @@
 
 ### Enhancements
 
+* Added `RealmConfiguration.Builder.compactOn()` to compact the file on launch (#3739).
```
CompactOnLaunch?

dalinaum replied:
Thanks.

zaki50 commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -672,6 +692,19 @@ public Builder readOnly() {
+         *                        the total file size (data + free space) and the total bytes used by data in the file.
+         */
+        public Builder compactOnLaunch(CompactOnLaunchCallback compactOnLaunch) {
+            this.compactOnLaunch = compactOnLaunch;
```
I think we should have a check if `compactOnLaunch` is not null.
If users pass `null` as callback, current implementation does not compact file at all.

dalinaum replied:
OK. I will add a check.

cmelchior commented on realm/realm-library/src/main/java/io/realm/CompactOnLaunchCallback.java:
```diff
@@ -0,0 +1,35 @@
+package io.realm;
+
+/**
+ * CompactOnLaunchCallback interface is used to determine if it should be compacted before being returned to the user.
```
I would rewrite it to this:

```
This interface is used to determine if a Realm file should be compacted the first time the file is opened and before the instance is returned.
<p>
Note that compacting a file can take a while, so compacting should generally only be done on a background thread or when used in combination with {@link Realm#getInstanceAsync(RealmConfiguration, Callback)}.
```

cmelchior commented on realm/realm-library/src/main/java/io/realm/CompactOnLaunchCallback.java:
```diff
@@ -0,0 +1,35 @@
+public interface CompactOnLaunchCallback {
+
+    /**
+     * This method determines if it should be compacted before returned to the user. It is passed the total file size
```
if it -> if the Realm file should be compacted before being opened and returned to the user.

kneth replied:
"if it" -> "if the Realm file"

cmelchior commented on realm/realm-library/src/main/java/io/realm/CompactOnLaunchCallback.java:
```diff
@@ -0,0 +1,35 @@
+     * This method determines if it should be compacted before returned to the user. It is passed the total file size
+     * (data + free space) and the total bytes used by data in the file.
+     *
+     * @param totalBytes the total file size (data + free space)
```
Missing .

cmelchior commented on realm/realm-library/src/main/java/io/realm/CompactOnLaunchCallback.java:
```diff
@@ -0,0 +1,35 @@
+     * (data + free space) and the total bytes used by data in the file.
+     *
+     * @param totalBytes the total file size (data + free space)
+     * @param usedBytes the total bytes used by data in the file
```
Missing .

cmelchior commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -194,6 +197,17 @@ String getAssetFilePath() {
     }
 
     /**
+     * Returns a callback to determine if it should be compacted before being returned to the user.
```
Returns a callback to determine if the Realm file should be compacted

cmelchior commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -194,6 +197,17 @@ String getAssetFilePath() {
+     * it should be compacted before being returned to the user. It is passed the total file size (data + free space)
+     * and the total bytes used by data in the file.
+     */
+    public CompactOnLaunchCallback getCompactOnLaunch() {
```
`getCompactOnLaunchCallback()` looks more readable to me?

cmelchior commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -672,6 +692,22 @@ public Builder readOnly() {
         }
 
+        /**
+         * Sets compactOnLaunch to determine if it should be compacted before returned to the user. It is passed the
```
Set this to determine if the Realm file should be

kneth replied:
"if it" -> "if the Realm file"

kneth commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1056,6 +1057,44 @@ public void compactRealm_onExternalStorage() {
+        Pair<Long, Long> results = populateTestRealmAndCompactOnLaunch(new CompactOnLaunchCallback() {
+            @Override
+            public boolean shouldCompact(long totalBytes, long usedBytes) {
+                assertTrue(totalBytes > usedBytes);
```
`>=` seems be bit more relaxed.

beeender replied:
I think you can create 100 objects and then remove them. Those 100 deleted objects will still take space in the file until compacting. So there would be obvious results after compacting called. Then it should be always true here. (until GC is implemented in core.)

cmelchior replied:
Good idea 👍

dalinaum replied:
Added deletion code that @beeender suggested.

dalinaum replied:
I adopted @beeender 's idea. Since `populateTestRealm` makes many objects, I added only `realm.deleteAll()`.

cmelchior commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -672,6 +692,22 @@ public Builder readOnly() {
+         *                        to determine if it should be compacted before being returned to the user. It is passed
+         *                        the total file size (data + free space) and the total bytes used by data in the file.
+         */
+        public Builder compactOnLaunch(CompactOnLaunchCallback compactOnLaunch) {
```
IMO we should also provide an easy to use configuration option with just a `compactOnLaunch()` override as well. Then we can provide our own default callback implementation `DefaultCompactOnLaunchCallback()`  with semantics provided by us.

I don't think that most users are in a better position that us to provide good heuristics around this, so we might as well make it really easy to use. 

Good starting heuristics could perhaps just be those used here:  https://news.realm.io/news/realm-objc-swift-2.6 although 100 MB feels a bit much on Android.

kneth commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -194,6 +197,17 @@ String getAssetFilePath() {
+     *
+     * @return a callback called when opening a Realm for the first time during the life of a process to determine if
+     * it should be compacted before being returned to the user. It is passed the total file size (data + free space)
+     * and the total bytes used by data in the file.
```
remove "total"

dalinaum commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1056,6 +1057,44 @@ public void compactRealm_onExternalStorage() {
+                return true;
+            }
+        });
+        assertTrue(results.first.compareTo(results.second) > 0);
```
Does `result.first > results.second` compare their reference values? Should it be `assertTrue(results.first.longValue() > results.second.longValue())`?

cmelchior replied:
`Long > Long` works the same way `long > long` due to auto-unboxing https://docs.oracle.com/javase/tutorial/java/data/autoboxing.html

cmelchior commented on realm/realm-library/src/main/java/io/realm/CompactOnLaunchCallback.java:
```diff
@@ -0,0 +1,35 @@
+ * limitations under the License.
+ */
+
+package io.realm;
```
I think the primary reason for Transaction to be an inner class is because we have two, one for DynamicRealm and one for Realm. 

You could argue it could have been an outer class like: 
```
public interface Transaction<T extends BaseRealm> {
   void execute(T realm);
}
```
But that would have exposed the concept of the `BaseRealm` in the public API.
But generally, I would tend to agree that outer classes look slightly better.

Inner classes have the benefit that they scope the interface to a very specific use case though.

IMO, In this case, it seems most natural to keep it as a separate interface

dalinaum commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1056,6 +1057,44 @@ public void compactRealm_onExternalStorage() {
+    }
+
+    @Test
+    public void compactOnLaunch_shouldCompact() throws IOException {
```
Should it be a separated PR?

cmelchior replied:
No, I think it would be good to have it in this one.

dalinaum replied:
Added a test as the following:
https://github.com/realm/realm-java/pull/4857/commits/5244336c2241098b2c1b8dc205d436bfbe377e37#diff-eb98b3546f9b692ff01aaff903a3bdb3R1118

dalinaum commented on realm/realm-library/src/androidTest/java/io/realm/rule/TestRealmConfigurationFactory.java:
```diff
@@ -129,7 +125,16 @@ public RealmConfiguration createConfiguration(String name, Object module) {
         return createConfiguration(null, name, module, null);
     }
 
+    public RealmConfiguration createConfiguration(String name, CompactOnLaunchCallback compactOnLaunchCallback) {
```
I agree on that point, I will remove `createConfiguration(String, String, Object, byte[], CompactOnLaunchCallback)`.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1056,6 +1057,52 @@ public void compactRealm_onExternalStorage() {
+        realm.close();
+        long before = new File(realmConfig.getPath()).length();
+        RealmConfiguration.Builder builder = configFactory.createConfigurationBuilder();
+        File folder = configFactory.getRoot();
```
Please use the same pattern we do elsewhere. Also, this is already being set in the configFactory, so

```
RealmConfiguration realmConfig = configFactory.createConfigurationBuilder()
     .name(REALM_NAME)
     .compactOnLaunch(compactOnLaunch)
     .build();
```

dalinaum commented:
I updated this PR as reviews, added more tests, fixed a bug and added a simpler simpler `builder.compactOnLaunch()`.

Please review this again @realm/java

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1056,6 +1059,137 @@ public void compactRealm_onExternalStorage() {
+                bgRealm.close();
+                bgThreadDoneLatch.countDown();
+            }
+        }).run();
```
This is not starting another thread, but running the runnable on current thread. You need to use `start()` and `thread.join()` to wait for it to finish

cmelchior commented on realm/realm-library/src/main/java/io/realm/DefaultCompactOnLaunchCallback.java:
```diff
@@ -0,0 +1,30 @@
+import io.realm.log.RealmLog;
+
+/**
+ * The CompactOnLaunchCallback that checks whether the Realm file size exceeds 50 MB and its used bytes are less than 50%.
```
I would rephrase this to:
```
The default implementation for determining if a file should be compacted or not. This implementation will only trigger if the file is above 50 MB and more than 50% can be reclaimed.
```

cmelchior commented on realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:
```diff
@@ -672,6 +692,30 @@ public Builder readOnly() {
         }
 
+        /**
+         * Setting this will cause the Realm to compact the Realm file before returned to the user if the Realm file
```
Don't mention specific details here as they might well change, and we will surely forget to update this. Perhaps something like:

```
Setting this will cause Realm to compact the Realm file if the Realm file has grown too large and a significant amount of space can be recovered. See {@link DefaultCompactOnLaunchCallback} for details.
``

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1056,6 +1059,147 @@ public void compactRealm_onExternalStorage() {
+
+        realm.close();
+        try {
+            thread.join();
```
Why join twice?

dalinaum replied:
Ups, I had a stupid mistake again while asleep. I should be careful more.
